src/compiler/checker.ts

```diff
@@ -8964,6 +8964,13 @@ namespace ts {
                         (isTypeSubsetOf(globalObjectType, target) || (!isComparingJsxAttributes && isEmptyObjectType(target)))) {
                         return false;
                     }
+                    if (target.flags & TypeFlags.Union) {
+                        const discriminantType = findMatchingDiscriminantType(source, target as UnionType);
+                        if (discriminantType)  {
+                            // check excess properties against discriminant type only, not the entire union
+                            return hasExcessProperties(source, discriminantType, reportErrors);
+                        }
+                    }

```

Comment:
`{ tag: "A", x } | { tag: "A", y }` already works with the current PR:

```ts
type AAA = { tag: "A", x: number } | { tag: "A", y: string };
let aaa: AAA;
aaa = { tag: "A", x: 12 }; // ok
aaa = { tag: "A", y: "Hi" }; // ok
aaa = { tag: "A", x: 12, extra: "extra" } // error
```

So I'm not sure how to construct a non-unique discriminant example that fails.